src/eth_provider/contracts/erc20.rs

```diff
@@ -34,17 +31,29 @@ impl<P: EthereumProvider> EthereumErc20<P> {
     }
 
     pub async fn balance_of(self, evm_address: Address, block_id: BlockId) -> EthProviderResult<U256> {
-        // Prepare the calldata for the bytecode function call
-        let address = EthersAddress::from_slice(evm_address.as_slice());
-        let calldata = IERC20Calls::BalanceOf(BalanceOfCall { account: address }).encode();
+        // Prepare the calldata for the bytecode function call.
+        // Create a provider using the `ProviderBuilder`` and the RPC configuration from the environment.
+        let provider = ProviderBuilder::new()
+            .on_builtin(&RPCConfig::from_env().expect("Failed to load Kakarot RPC config").socket_addr)
+            .await
+            .expect("Failed to create provider via alloy `ProviderBuilder`");
```

Comment:
Is there no way to get the calldata without using a false provider?

Reply:
Fixed, I found a much better solution with simply `abi_encode`